binding/rust/src/util.rs

```diff
@@ -0,0 +1,167 @@
+        machine if RPI_MACHINES.contains(&machine.as_str()) => {
+            if cfg!(target_arch = "aarch64") {
+                PathBuf::from(format!(
+                    "lib/raspberry-pi/{}-aarch64/libpv_porcupine.so",
+                    machine
+                ))
+            } else {
+                PathBuf::from(format!("lib/raspberry-pi/{}/libpv_porcupine.so", machine))
+            }
+        }
+        machine if JETSON_MACHINES.contains(&machine.as_str()) => {
+            PathBuf::from("lib/jetson/cortex-a57-aarch64/libpv_porcupine.so")
+        }
+        "beaglebone" => PathBuf::from("lib/beaglebone/libpv_porcupine.so"),
+        _ => {
+            warn!("WARNING: Please be advised that this device (CPU part = {}) is not officially supported by Picovoice.\nFalling back to the armv6-based (Raspberry Pi Zero) library. This is not tested nor optimal.\nFor the model, use Raspberry Pi's models", cpu_part);
+            PathBuf::from("lib/raspberry-pi/arm11/libpv_porcupine.so")
+        }
+    };
+}
+
+pub fn pv_library_path<P: AsRef<Path>>(relative: P) -> PathBuf {
+    let mut path = PathBuf::from(relative.as_ref());
+    path.push(base_library_path());
+    return path;
+}
+
+pub fn pv_model_path<P: AsRef<Path>>(relative: P) -> PathBuf {
+    let mut path = PathBuf::from(relative.as_ref());
+    path.push("lib/common/porcupine_params.pv");
+    return path;
+}
+
+#[cfg(target_os = "macos")]
+fn keyword_path_subdir() -> PathBuf {
+    return PathBuf::from("mac");
+}
+
+#[cfg(target_os = "windows")]
+fn keyword_path_subdir() -> PathBuf {
+    return PathBuf::from("windows");
+}
+
+#[cfg(all(target_os = "linux", target_arch = "x86_64"))]
+fn keyword_path_subdir() -> PathBuf {
+    return PathBuf::from("linux");
+}
+
+#[cfg(all(target_os = "linux", any(target_arch = "arm", target_arch = "aarch64")))]
+fn keyword_path_subdir() -> PathBuf {
+    let machine = find_machine_type();
+    return match machine {
+        machine if RPI_MACHINES.contains(&machine.as_str()) => PathBuf::from("raspberry-pi"),
+        machine if JETSON_MACHINES.contains(&machine.as_str()) => PathBuf::from("jetson"),
+        "beaglebone" => PathBuf::from("beaglebone"),
+        _ => {
+            panic!("ERROR: Please be advised that this device (CPU part = {}) is not officially supported by Picovoice.");
+        }
+    };
+}
+
+pub fn pv_keyword_paths<P: AsRef<Path>>(relative: P) -> HashMap<String, String> {
+    let mut dir = PathBuf::from(relative.as_ref());
+    dir.push("resources/keyword_files/");
+    dir.push(keyword_path_subdir());
+
+    let mut keyword_paths = HashMap::new();
+    let dir_entries = fs::read_dir(dir).unwrap();
+    for entry in dir_entries {
+        let entry = entry.unwrap();
+        let path = entry.path();
+        let keyword = entry
+            .file_name()
+            .into_string()
+            .unwrap()
+            .split("_")
+            .next()
+            .unwrap()
+            .to_string();
```

Comment:
can this fail? if so, we should probably check if the entry conforms to our file name pattern and exists before adding it